### PR TITLE
Execute preinstall before tree is calculated

### LINKED ISSUE #10379: preinstall execution order in npm@3.x
Sorry if this has been brought up before, but it looks like the preinstall step for my local repo now runs _after_ dependent modules have been installed with `npm@3.x`?

With `npm@2.x` we used the preinstall hook to delete our `node_modules` folder & clone it from a git repo...so then the `install` step would execute, see that everything was already installed, and do nothing.  With `npm@3.x` we have the same `preinstall` hook, but first npm does an install into the `node_modules` folder itself _then_ it runs the `preinstall` hook.  Is this a bug?  It seems like its not really a "pre-install" step anymore if it actually fires _after_ the install...

### PR DESCRIPTION
**Pull request by palmerj3**
This PR attempts to fix https://github.com/npm/npm/issues/10379 by executing preinstall before the tree is calculated and an ideal tree generated.

Functionally this does appear to work but I'm looking for feedback and testing.

@iarna @zkat would love your feedback on this if you have time. Thanks!

### CODE REVIEW

File: lib/install.js [hunk 1]
```diff
@@ -208,7 +208,7 @@ function Installer (where, dryrun, args) {
   this.todo = []
   this.progress = {}
   this.noPackageJsonOk = !!args.length
-  this.topLevelLifecycles = !args.length
+  this.topLevelLifecycles = !!args.length
```
Comment (legodude17) [inline lib/install.js]: Just wondering, why did you change this?

Comment (palmerj3) [inline lib/install.js]: Because preinstall was never being called at all without this and I think it was a bug to begin with given how other args.length are being handled. This is the code in reference that was always triggered with args.length being the way it was https://github.com/palmerj3/npm/blob/0ee8d4248b4294c347923b8cff41f89dfe597ac9/lib/install.js#L541

**Comment by zkat** [comment]
@palmerj3 thanks a lot for tackling this! Ping us when it's ready! I'm gonna /cc @iarna cause she made the original call to move this up. :)

We have a nice window for a breaking change like this coming up (and I'm pretty sure it's a breaking change -- correct me if I'm wrong), so if you get it done in the next ~2 weeks, we'd be able to land it as part of `npm@5`. It might be longer than that, because that depends on other factors, but hopefully it's still incentive :)

File: lib/install.js (hunk 1 above)
Comment (iarna) [inline lib/install.js]: Your `preinstall` top level lifecycle shouldn't run if you run `npm install foobar`. It should only run if you're installing the top level package, eg `npm install`.

Comment (palmerj3) [inline lib/install.js]: @iarna ah ok good point :)

File: lib/install.js [hunk 2]
```diff
@@ -242,8 +242,15 @@ Installer.prototype.run = function (_cb) {
     }
   }
 
+  var preinstallSteps = []
   var installSteps = []
   var postInstallSteps = []
+
+  preinstallSteps.push(
```
Comment (iarna) [inline lib/install.js]: Why introduce a new phase here? Why not just make this the start of the install steps? Either way, this needs to not run of `this.dryrun` is true.

Comment (palmerj3) [inline lib/install.js]: @iarna A new phase is optional and I'm fine removing it. Before getting this to work I had a few other steps there and it made some sense. Removing now and pushing a few tweaks!

Comment (palmerj3) [inline lib/install.js]: Actually let me edit that... it does make some sense because, in the case where you want to bail before install, then I think having it organized this way will allow for this whereas having everything chained in a single install step would mean you'd wait longer for that feedback?

**Comment by palmerj3** [comment]
@zkat thanks! :)  I'll work on the tweaks mentioned and write a few tests then get your feedback again.

**Review by iarna — CHANGES_REQUESTED** [review]
With a couple of tweaks I think we should be able to take this.

**Comment by iarna** [comment]
Other work that will need to be done:

* [ ] Travis should be green (all tests passing)
* [ ] It would be nice to have a test that can only pass if `preinstall` runs early.

**Comment by palmerj3** [comment]
Ok cool! I'm working on making the tests green now.

**Comment by palmerj3** [comment]
@iarna @zkat I got the tests to pass. If you get some time please review.. the main complication is the use of [tacks](https://github.com/iarna/tacks) in the tests in order to generate fake package.json files. So this latest iteration handles that. I would much prefer to not have a code path just for tests so if you have any suggestions I would love to hear.

In the meantime I'll try to work on a test that will only pass if preinstall is run early.

**Comment by palmerj3** [comment]
@zkat I'm not sure this would be a breaking change... I'll totally leave that up to you all to decide. Really the only way this would be breaking is if devs attempted to manipulate the tree preinstall, it failed, and they are now relying on that :)

**Comment by selvaa89** [comment]
Any idea when this PR will be merged? Our repos mutates package.json on preinstall script for various reasons. Because of this issue we are stuck with 2.x!

**Comment by palmerj3** [comment]
@zkat @iarna looking at the other tests I don't see any sort of stubbing libs being used like sinon.. so I'm not totally sure how I could reasonably write a test for this if I can't spy on the "loadCurrentTree" method. Thoughts?

**Comment by iarna** [comment]
@palmerj3 We stub with `requireInject`, but I would be inclined to go with a functional test for this. For example, you could:

1. Pass in a registry that can't work. (eg, `https://127.0.0.11:1111`) 
2. Have a `preinstall` that moves a dependency into `node_modules`
3. Try to run the install. It should succeed because its dependency should be in place due to the preinstall. If the new behavior is not in place then it will fail with a network error.

**Comment by palmerj3** [comment]
@iarna thank you! I will take that approach.

**Comment by palmerj3** [comment]
@iarna I took your advice pretty much verbatim and wrote a test that fails on "latest" and passes with this PR to verify preinstall fires early.

Thanks again for the advice and direction!

File: test/tap/install-parse-error.js [hunk 3]
```diff
@@ -39,7 +39,6 @@ test('failing to parse package.json should be error', function (t) {
     function (err, code, stdout, stderr) {
       if (err) throw err
       t.equal(code, 1, 'exit not ok')
-      t.similar(stderr, /npm ERR! Failed to parse json/)
```
Comment (legodude17) [inline test/tap/install-parse-error.js]: Why was this line removed?

File: scripts/changelog.js [hunk 4]
```diff
@@ -75,7 +75,6 @@ function main () {
     if (/^---$/.test(line)) {
       print_commit(commit)
     } else if (m = line.match(/^([a-f0-9]{7,9}) ([a-f0-9]+) (?:[(]([^)]+)[)] )?(.*?) [(](.*?)[)]/)) {
-
```
Comment (legodude17) [inline scripts/changelog.js]: Any idea why this was changed?

**Review by legodude17 — COMMENTED** [review]
😄

File: test/tap/install-parse-error.js (hunk 3 above)
Comment (palmerj3) [inline test/tap/install-parse-error.js]: Because I wasn't getting the same error message. Was an error none-the-less but not with that specific text. And I don't think this is a valuable assertion to have and has the risk of breaking across node versions.

File: scripts/changelog.js (hunk 4 above)
Comment (palmerj3) [inline scripts/changelog.js]: I only changed it because the linter was complaining about it

File: test/tap/install-parse-error.js (hunk 3 above)
Comment (legodude17) [inline test/tap/install-parse-error.js]: Ok, fair enough.

File: scripts/changelog.js (hunk 4 above)
Comment (legodude17) [inline scripts/changelog.js]: I am surprised the linter complained about code that was already there.

Comment (palmerj3) [inline scripts/changelog.js]: Someone else sent a fix for the same issue but it hasn't been merged. - https://github.com/npm/npm/pull/16110

File: lib/install.js [hunk 5]
```diff
@@ -536,8 +545,21 @@ Installer.prototype.runPreinstallTopLevelLifecycles = function (cb) {
...
+  }
+
+  // Construct initial node for top of the tree
+  var rootNode = {
+    package: p
+  }
+
   steps.push(
-    [doOneAction, 'preinstall', this.idealTree.path, this.idealTree, trackLifecycle.newGroup('preinstall:.')]
+    [doOneAction, 'preinstall', process.cwd(), rootNode, trackLifecycle.newGroup('preinstall:.')]
```
Comment (iarna) [inline lib/install.js]: `process.cwd()` isn't safe for this, as you can run `npm install` in a subdir of a project and it'll install at the project root.

File: test/tap/install-parse-error.js (hunk 3 above)
Comment (iarna) [inline test/tap/install-parse-error.js]: This assertion needs to stand. The error message it's checking for is from npm and so Node.js version doesn't matter.

If this assertion is failing it means that something is unexpected about the error message. The install failed but what error did it report?

File: scripts/changelog.js (hunk 4 above)
Comment (iarna) [inline scripts/changelog.js]: It has as of last Friday—if you rebase on to the latest `release-next` you should see it go away.

File: lib/install.js (hunk 2 above)
Comment (iarna) [inline lib/install.js]: Chains are run in sequence and stop after the first error. The only reason to have multiple chains is if you want to run a later chain even though an earlier one failed.

File: lib/install.js [hunk 6]
```diff
@@ -242,8 +242,17 @@ Installer.prototype.run = function (_cb) {
     }
   }
 
+  var preinstallSteps = []
```
Comment (iarna) [inline lib/install.js]: https://github.com/npm/npm/pull/16013#discussion_r108286379

File: lib/install.js (hunk 5 above)
Comment (iarna) [inline lib/install.js]: What you want here is `this.where` I believe.

File: scripts/changelog.js (hunk 4 above)
Comment (palmerj3) [inline scripts/changelog.js]: @iarna this PR is toward "latest" - should I switch to release-next?

File: lib/install.js (hunk 6 above)
Comment (palmerj3) [inline lib/install.js]: Ok cool I'll strip that out.

**Comment by palmerj3** [comment]
Thank you again @iarna for the guidance! I'll make the changes requested and look further into the json error.

**Comment by xkr47** [comment]
@palmerj3 ping?

**Comment by palmerj3** [comment]
I haven't had time recently to make the final changes but I hope to soon. I wish I had received feedback all at once instead of in two phases but alas changes are required still and I hope to get to them soon.

**Comment by zkat** [comment]
This was done as part of `npm@5.0.0`. In the rush, we shoved in our own code, but thank you for filing this PR, and I hope `npm@5` works like you need it to, now 🎉
